test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/PreserveTypesAndAttributesTests.cs

```diff
@@ -0,0 +1,395 @@
+            for (var i = 0; i < result1.Count; i++)
+            {
+                Assert.Equal(result3[i].Id, result2[i].Id);
+                Assert.Equal(result3[i].CorporationName, result2[i].CorporationName);
+                Assert.Equal(result3[i].City, result2[i].City);
+                Assert.Equal(result3[i].Name, result2[i].Name);
+            }
+        }
+    }
+
+
+    // One level of inheritance classes
+
+    public interface ICustomer
+    {
+        string Id { get; set; }
+        string City { get; set; }
+        string CorporationName { get; set; }
+        string Name { get; set; }
+    }
+
+    [Key("CustomerID")]
+    public class OneLevelCustomer1 : ICustomer
+    {
+        [OriginalName("CustomerID")]
+        public string Id { get; set; }
+
+        public string City { get; set; }
+
+        [OriginalName("CompanyName")]
+        public string CorporationName { get; set; }
+
+        [OriginalName("ContactName")]
+        public string Name { get; set; }
+    }
+
+    [Key("CustomerID")]
+    public class BaseCustomer
+    {
+        [OriginalName("CustomerID")]
+        public string Id { get; set; }
+    }
+
+    public class OneLevelCustomer2 : BaseCustomer
+    {
+        public string City { get; set; }
+
+        [OriginalName("CompanyName")]
+        public string CorporationName { get; set; }
+
+        [OriginalName("ContactName")]
+        public string Name { get; set; }
+    }
+
+
+    // More than one level of inheritance classes
+
+    public abstract class ACustomer : ICustomer
+    {
+        public virtual string Id { get; set; }
+        public virtual string City { get; set; }
+        public virtual string CorporationName { get; set; }
+        public virtual string Name { get; set; }
+    }
+
+    [Key("CustomerID")]
+    public class TwoLevelCustomer1 : ACustomer
+    {
+        [OriginalName("CustomerID")]
+        public override string Id { get; set; }
+
+        public override string City { get; set; }
+
+        [OriginalName("CompanyName")]
+        public override string CorporationName { get; set; }
+
+        [OriginalName("ContactName")]
+        public override string Name { get; set; }
+    }
+
+    public interface ICustomer2 : ICustomer
+    {
+        List<Order> OrdersList { get; set; }
+    }
+
+    [Key("CustomerID")]
+    public class TwoLevelCustomer2 : ICustomer2
+    {
+        [OriginalName("CustomerID")]
+        public string Id { get; set; }
+
+        public string City { get; set; }
+
+        [OriginalName("CompanyName")]
+        public string CorporationName { get; set; }
+
+        [OriginalName("ContactName")]
+        public string Name { get; set; }
+
+        [OriginalName("Orders")]
+        public List<Order> OrdersList { get; set; }
+    }
+
+    [Key("OrderID")]
+    public class Order
+    {
+        public int OrderID { get; set; }
+        public decimal Freight { get; set; }
+    }
+
+
+    public class TestContext
```

Comment:
The name `TestContext` is very generic, but this class is very specific to this test. Already, this will cause conflicts and compilation errors if this other PR https://github.com/OData/odata.net/pull/2863 is merged, because they both define a `TestContext` class in the same namespace. One way to avoid this conflict is to move `TestContext` as an inner class of the test class. But if we do that, it brings me back to the question, why do we need this class in the first place?

The helper methods it provides might as well be moved to the test class then, then we don't need an inner class. We can move them to the bottom of the class so that they don't create distractions or noise when looking at tests.